### DISCUSSION TITLE
Seizure detector

### DISCUSSION THREAD

**Discussion by Spiralo-Idioide** [InfiniTimeOrg/InfiniTime#1800] [Ideas]
Hello everyone !

# The main idea

I have quite frequent seizures and my family and friends are anxious that I might have one alone or somewhere where the people around me don't know how to react.
So I had the idea to use the pinetime as a seizure detector that connects to an android phone in order to send a message when I might have a seizure. 

The kind of seizures that might be possible to detect would be :
- Clonic 
- Tonic-clonic 
- Focal to bilateral tonic-clonic

I wanted to share with the pinetime community this idea to know if it would be possible to implement. 
Also I am sorry if my english is not perfect, if anything is hard to understand, don't hesitate to ask me, I will try to explain it better.
Also I mostly say _"I"_ but it should be understand as _"any owner that uses this project"_

# Why the pinetime ?

- Because it is cheap. Seizure detector devices exist but are usually quite expensive and often necessitate an also quite expensive subscription.
- It is open-source and I like open-source. Also it would make a lot easier for anyone to improve the detector.
- I have wanted to learn C++ for some time and always gave up. This project might be the thing that I need

# Possible functions

## Main functions
- The seizure detector, running on the watch (see '# How I think it might work')
  - Make the detector activable / de-activable in settings so it won't bother people that don't need it (most people)
- Make the watch vibrate after 10 to 30 second of possible seizure _(configurable on the watch)_ to allow false positives to be discarded by the user
  - If the user does not discard the message on the watch for 10 to 60 seconds _(configurable on the watch)_ then the detector considered a seizure happened and will make the watch and phones do what they are configured to do in case of a detected seizure
- Store last 12 to 72 hours of data _(configurable on the watch)_ so it is possible for the user to send it for analysis in order to improve the detector (especialy for false positives)
  - Store detected seizures data for a longer time (maybe a week) so the user have more time to send it (add sometihing if the detected seizure is a false positive)
- Connect with an android app that can send a message to a selection of contacts
  - I don't think the current android apps that connect to the pinetime can send messages 
  - So I thought of Tasker, wich is a paid app but is very cheap and doesn't need a subscription
- Any other idea is welcome !

## Possible other functions
Everything here should be possible to activate/deactivate directly with tasker so it is not needed to program it into the pinetime

- Send location data with the messages _(possible with tasker)_
- Detect the possible end of the seizure
  - Send a message if the detector think the seizure is finished
  - Send a message every 30 to 180s _(configurable on the watch)_ informing that the seizure is still ongoing
- Play an audio file on the phone with instructions on what to do _(probably possible with tasker)_
  - Maybe put a message with the instructions on the phone _(might be possible with tasker)_
- Change the notifications settings of the phone to something that would make a lot of noise so I would be sure that I can hear a possible call when I regain consciousness _(probably possible with tasker)_
- Allow the user to send messages directly from the watch about the seizure _(possible with tasker)_
  - A message to inform that I regained consciousness 
  - Re-send the infos about location data
  - A message in case the detector detected a false positive to inform that it was in fact a false positive
  - A message in case the detector didn't detect a seizure so when I regain consciousness I can inform my contacts of the seizure
  - A message to inform my contacts that I have important pain/wounds and I need urgent help
  - A message to ask my contacts if they can come to pick me up (with location data) 
  - Send a new message after 10 to 60mn _(configurable on the watch)_ without interaction with the watch/phone
- Show a notification (watch and/or phone) 6 to 48 hours after a detected seizure to ask the user if he would like to send the data about the seizure. If he accepts, then show the instructions on how to do it (because I am not sure it is possible to send the data directly with the watch/phone)
- Any other idea is welcome !

# How I think it might work

## Detector

The detector would mainly work using accelerometer data. 
Heart rate data could be useful but would most likely be hard to interpret without a lot of seizures data.
Every 1 to 5s an analysis is done on the data to detect a possible seizure.

### Accelerometer
**Help/ideas are needed here**

Last 30 to 60s of accelerometer data would be stored in the ram with interval of 0.05 to 0.5s. 

The analysis would most likely try to interpret the acceleration as movement on the axis. If a lot of fast movements are detected on the 3 axis at the same time, then the seizure detector considers a seizure is detected.
At least some data will be needed to have as little false positives/negatives.

## Storage
**Help/ideas are needed here**

Storage will work with files stored every 5 to 60mn. The name of the file will be the time where the file data begins (rename it if a seizure is detected during the time range it covers).
If possible the best would be to store the data on the user phone instead of on the watch.

### File size for accelerometer

I am not sure how to store raw accelerometer data without having file sizes impossible to store in the watch memory which is very limited.
Maybe it might be possible to lossy compress the data on 2 to 4 bytes per axis and not storing the data every 0.05 second.

**Min file size :**
2 bytes per axis, stored every 0.5s.
(3 x 2) x 7200 = 43.2 kB or 0.0432 MB per hour
1,0368MB to store the last 24h.

**Max file size :** 
4 bytes per axis stored every 0.05s.
(3 x 4) x 72000 = 864 kB or 0.864 MB  per hour _(definitely impossible to store on the watch)_

### File size for heart rate
Because heart rate data doesn't need high precision, storing the average of the last 5 to 10s will not be a problem at all. It is even possible to store as 2 bytes, HR over 255 being very rare.

File size would be 0.72kB to 1.44kB per hour. 
17,28kB to 34,56kB to store the last 24h.

# Conclusion

I have no experience with the pinetime (I don't even own one at time of writing) and because of that it is possible that 95% of my ideas are materially impossible. Maybe the core idea in itself is entirely impossible to achieve.

Anyway, thank you to everyone that will read this post. And thank you to everyone that will take part on the discussion, even if it is to explain why this project is impossible to achieve.

And thanks again !

**Comment by SteveAmor** [comment]
This may help. Not sure what the progress is with Pinetime.  https://www.openseizuredetector.org.uk/?page_id=1480#PineTime

**Comment by jones139** [comment]
Excellent idea.  This is very close to what we have with OpenSeizureDetector.   There are currently two 'nearly working ' implementations for pinetime (we do all the processing on the phone so the watch app can be quite simple).   If you have a look here: https://github.com/orgs/OpenSeizureDetector/discussions/70, I can talk you through what would be involved in getting it working.
